src/Mod/Path/PathScripts/PathVcarve.py

```diff
@@ -0,0 +1,286 @@
+# ***************************************************************************
+
+import FreeCAD
+import Part
+import Path
+import PathScripts.PathEngraveBase as PathEngraveBase
+import PathScripts.PathLog as PathLog
+import PathScripts.PathOp as PathOp
+import PathScripts.PathUtils as PathUtils
+import PathScripts.PathGeom as PathGeom
+import PathScripts.PathPreferences as PathPreferences
+
+import traceback
+
+import math
+
+from PySide import QtCore
+
+__doc__ = "Class and implementation of Path Vcarve operation"
+
+
+if False:
+    PathLog.setLevel(PathLog.Level.DEBUG, PathLog.thisModule())
+    PathLog.trackModule(PathLog.thisModule())
+else:
+    PathLog.setLevel(PathLog.Level.INFO, PathLog.thisModule())
+
+
+# Qt tanslation handling
+def translate(context, text, disambig=None):
+    return QtCore.QCoreApplication.translate(context, text, disambig)
+
+
+class ObjectVcarve(PathEngraveBase.ObjectOp):
+    '''Proxy class for Vcarve operation.'''
+
+    def opFeatures(self, obj):
+        '''opFeatures(obj) ... return all standard features and edges based geomtries'''
+        return PathOp.FeatureTool | PathOp.FeatureHeights | PathOp.FeatureBaseFaces
+
+    def setupAdditionalProperties(self, obj):
+        if not hasattr(obj, 'BaseShapes'):
+            obj.addProperty("App::PropertyLinkList", "BaseShapes", "Path",
+                            QtCore.QT_TRANSLATE_NOOP("PathVcarve",
+                                "Additional base objects to be engraved"))
+        obj.setEditorMode('BaseShapes', 2)  # hide
+        if not hasattr(obj, 'BaseObject'):
+            obj.addProperty("App::PropertyLink", "BaseObject", "Path",
+                            QtCore.QT_TRANSLATE_NOOP("PathVcarve",
+                            "Additional base objects to be engraved"))
+        obj.setEditorMode('BaseObject', 2)  # hide
+
+    def initOperation(self, obj):
+        '''initOperation(obj) ... create vcarve specific properties.'''
+        obj.addProperty("App::PropertyFloat", "Discretize", "Path",
+                        QtCore.QT_TRANSLATE_NOOP("PathVcarve",
+                        "The deflection value for discretizing arcs"))
+        obj.addProperty("App::PropertyFloat", "Threshold", "Path",
+                        QtCore.QT_TRANSLATE_NOOP("PathVcarve",
+                        "cutoff for removing colinear segments (degrees). \
+                        default=10.0."))
+        obj.addProperty("App::PropertyFloat", "Tolerance", "Path",
+                QtCore.QT_TRANSLATE_NOOP("PathVcarve", ""))
+        obj.Threshold = 10.0
+        obj.Discretize = 0.01
+        obj.Tolerance = PathPreferences.defaultGeometryTolerance()
+        self.setupAdditionalProperties(obj)
+
+    def opOnDocumentRestored(self, obj):
+        # upgrade ...
+        self.setupAdditionalProperties(obj)
+
+    def buildPathMedial(self, obj, Faces):
+        '''constructs a medial axis path using openvoronoi'''
+
+        def insert_many_wires(vd, wires):
```

Comment:
I suggest not defining functions inside functions, unless there is a very specific reason for that.

Reply:
fixed

Reply:
I'm going to leave this as is at least until the feature gets more testing.  If someone wants to refactor it, either for speed or to be more pythonic, that's fine.